src/adapter/src/rbac.rs

```diff
@@ -0,0 +1,236 @@
+    plan: &Plan,
+) -> Result<(), AdapterError> {
+    if !catalog.rbac_checks_enabled() {
+        return Ok(());
+    }
+
+    let role_id = session.role_id();
+    let Some(role) = catalog.try_get_role(role_id) else {
+        // PostgreSQL allows users that have their role dropped to perform some actions,
+        // such as `SET ROLE` and certain `SELECT` queries. We haven't implemented
+        // `SET ROLE` and feel it's safer to force to user to re-authenticate if their
+        // role is dropped.
+        return Err(AdapterError::ConcurrentRoleDrop(role_id.clone()));
+    };
+
+    if session.is_superuser() {
+        return Ok(());
+    }
+
+    if let Some(required_attribute) = generate_plan_attribute(plan) {
+        if !required_attribute.check_role(role) {
+            return Err(AdapterError::Unauthorized(UnauthorizedError::attribute(
+                plan.name().to_string(),
+                required_attribute,
+            )));
+        }
+    }
+
+    Ok(())
+}
+
+/// Generates the attributes required to execute a given plan.
+fn generate_plan_attribute(plan: &Plan) -> Option<Attribute> {
+    match plan {
+        Plan::CreateDatabase(_) => Some(Attribute::CreateDB),
+        Plan::CreateRole(_) | Plan::AlterRole(_) | Plan::DropRoles(_) => {
+            Some(Attribute::CreateRole)
+        }
+        Plan::CreateCluster(_) => Some(Attribute::CreateCluster),
+        Plan::CreateSource(_)
+        | Plan::CreateTable(_)
+        | Plan::CreateMaterializedView(_)
+        | Plan::CreateConnection(_)
+        | Plan::CreateSchema(_)
+        | Plan::CreateClusterReplica(_)
+        | Plan::CreateSecret(_)
+        | Plan::CreateSink(_)
+        | Plan::CreateView(_)
+        | Plan::CreateIndex(_)
+        | Plan::CreateType(_)
+        | Plan::DiscardTemp
+        | Plan::DiscardAll
+        | Plan::DropDatabase(_)
+        | Plan::DropSchema(_)
+        | Plan::DropClusters(_)
+        | Plan::DropClusterReplicas(_)
+        | Plan::DropItems(_)
+        | Plan::EmptyQuery
+        | Plan::ShowAllVariables
+        | Plan::ShowVariable(_)
+        | Plan::SetVariable(_)
+        | Plan::ResetVariable(_)
+        | Plan::StartTransaction(_)
+        | Plan::CommitTransaction
+        | Plan::AbortTransaction
+        | Plan::Peek(_)
+        | Plan::Subscribe(_)
+        | Plan::SendRows(_)
+        | Plan::CopyFrom(_)
+        | Plan::Explain(_)
+        | Plan::SendDiffs(_)
+        | Plan::Insert(_)
+        | Plan::AlterNoop(_)
+        | Plan::AlterIndexSetOptions(_)
+        | Plan::AlterIndexResetOptions(_)
+        | Plan::AlterSink(_)
+        | Plan::AlterSource(_)
+        | Plan::AlterItemRename(_)
+        | Plan::AlterSecret(_)
+        | Plan::AlterSystemSet(_)
+        | Plan::AlterSystemReset(_)
+        | Plan::AlterSystemResetAll(_)
+        | Plan::Declare(_)
+        | Plan::Fetch(_)
+        | Plan::Close(_)
+        | Plan::ReadThenWrite(_)
+        | Plan::Prepare(_)
+        | Plan::Execute(_)
```

Comment:
I'm a bit worried that this has the potential to escape the normal perms check. Can you add a test that verifies it doesn't incorrectly allow doing like a CREATE CLUSTER or something inside a DECLARE+EXECUTE?

Reply:
I made this change, but I think now the code inside `check_plan` is mostly dead code. I'm going to start at it for awhile and see if it can be cleaned up.